### PR TITLE
Introduce shellcheck for bash linting

### PR DESCRIPTION
**Pull request by aneeshdurg**
Shellcheck is a really nice sh linting tool. It's helped me catch some pretty nasty bugs before.

If no one has any objections to this PR the next steps would be linting existing bash scripts and then integrating this into our CI workflow.

In my experience shellcheck isn't super strict even with the style mode turned on, so I don't expect it being too much of a pain, and where necessary it's easy enough to disable it inline.

### CODE REVIEW

File: scripts/check_all.sh [hunk 1]
```diff
@@ -24,14 +24,19 @@ output_on_fail(){
...
-echo -ne 'running:  check_python_format.sh\r'
+display_running "check_python_format.sh"
 output_on_fail ${SCRIPT_DIR}/check_python_format.sh ${FIX} "${@}"
-#echo -ne 'running: check_python_lint.sh\r'
+#display_running "check_python_lint.sh"
 #output_on_fail pipenv run ${SCRIPT_DIR}/check_python_lint.sh "${@}"
-echo -ne 'Done!              \n'
+display_running "check_sh_format.sh"
+output_on_fail ${SCRIPT_DIR}/check_sh_format.sh ${FIX} "${@}"
+echo -e "${CLEAR_LINE}Done"
```
Comment (arthurp) [inline scripts/check_all.sh:42]: I didn't even know this existed. There are things I should add here. Ooops.

File: scripts/setup_ubuntu.sh [hunk 2]
```diff
@@ -85,7 +85,7 @@ run_as_original_user pip3 install --upgrade "pip$PIP_VERSION" "setuptools$SETUPT
 run_as_original_user pip3 install conan==1.33
 
 # Developer tools
-DEVELOPER_TOOLS="clang-format-10 clang-tidy-10 doxygen graphviz ccache cmake"
+DEVELOPER_TOOLS="clang-format-10 clang-tidy-10 doxygen graphviz ccache cmake shellcheck"
```
Comment (arthurp) [inline scripts/setup_ubuntu.sh:88]: Please add the dependency in `conda_recipe/environment.yml` and `scripts/setup_macos.sh` (if available in brew).

**Review by arthurp — APPROVED** [review]
LGTM. Checkers are good. My only comment is that the dependency on it is not added everywhere it needs to be. See the comment on `setup_ubuntu.sh`.

**Comment by arthurp** [comment]
The CI failure appears to be because the permissions on one of the scripts was accidentally changed.

**Comment by roshandathathri** [comment]
I'm in favor of this. Thanks, Aneesh!

**Pull request merged**